test/tests/functional/pbs_sched_badstate.py

```diff
@@ -0,0 +1,153 @@
+# coding: utf-8
+
+# Copyright (C) 1994-2019 Altair Engineering, Inc.
+# For more information, contact Altair at www.altair.com.
+#
+# This file is part of the PBS Professional ("PBS Pro") software.
+#
+# Open Source License Information:
+#
+# PBS Pro is free software. You can redistribute it and/or modify it under the
+# terms of the GNU Affero General Public License as published by the Free
+# Software Foundation, either version 3 of the License, or (at your option) any
+# later version.
+#
+# PBS Pro is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+# FOR A PARTICULAR PURPOSE.
+# See the GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+# Commercial License Information:
+#
+# For a copy of the commercial license terms and conditions,
+# go to: (http://www.pbspro.com/UserArea/agreement.html)
+# or contact the Altair Legal Department.
+#
+# Altair’s dual-license business model allows companies, individuals, and
+# organizations to create proprietary derivative works of PBS Pro and
+# distribute them - whether embedded or bundled with other software -
+# under a commercial license agreement.
+#
+# Use of Altair’s trademarks, including but not limited to "PBS™",
+# "PBS Professional®", and "PBS Pro™" and Altair’s logos is subject to Altair's
+# trademark licensing policies.
+
+from tests.functional import *
+
+
+class TestSchedBadstate(TestFunctional):
+
+    @timeout(600)
+    def test_sched_badstate_subjob(self):
+        """
+        This test case tests if scheduler goes into infinite loop
+        when following conditions are met.
+        - Kill a mom
+        - mark the mom's state as free
+        - submit an array job
+        - check the sched log for "Leaving sched cycle" from the time
+          array job was submitted.
+        If we are unable to find a log match then scheduler is in
+          endless loop and test case has failed.
+        """
+
+        self.mom.signal('-KILL')
+
+        attr = {'state': 'free', 'resources_available.ncpus': '2'}
+        self.server.manager(MGR_CMD_SET, NODE, attr, self.mom.shortname)
```

Comment:
I know that you just moved an existing test, but according to our documentation, 'state' cannot be set to 'free', so we shouldn't be setting it (as such, PBS should not even allow this).

Reply:
The bug for which test was written required a scenario where the server fails to send a subjob to a mom.... the way send job from server to mom could have failed is when mom is down but the node is seen as free by scheduler.
Therefore, even though the docs say we should not, in this case we have to set the node state to  'free'

Reply:
I understand, but it's not that it "shouldn't" be set, it's that it "cannot" be set. If it can be set today to "Free", then that's a bug, which might get fixed at some point and then this test will start failing. If this test was written to catch a race condition, then we should just remove it, race conditions are race conditions because they cannot be reproduced reliably, so no point writing a test to reproduce them.

Reply:
@agrawalravi90 there is nothing stopping you from setting the state of a node to free.  It is a valid node state.  It just isn't a good idea.  If a node is down and you set it back to free, it will just be marked down again when PBS realizes it is down.  I don't think this has ever been considered a bug, although we might want to disallow it as you suggest.  In any case, I think this test requires you to set the node to free while the mom is down.  That's what the docstring says anyway.  I think @Shrini-h wrote this test.  Maybe he can comment.

Reply:
Ok, I was going by the Ref guide which says that the permissions for the stat attr are 'r, r, r' when setting it to 'free'. I just don't want us to come back to this test and think about what to do with it if we decide to implement the documented behavior in the future.

Reply:
actually, I don't wanna hold this PR because of this test as it is an already existing test. If it causes issues in the future then we can come back to it.

Reply:
> I think @Shrini-h wrote this test. Maybe he can comment.

@bhroam   Sorry, I did not author this test. From git history, this has been introduced in PR https://github.com/PBSPro/pbspro/pull/264